### ISSUE TITLE
Support for generic workflow failure handler via interceptor

### ISSUE THREAD

**Issue by ddavidyuk** [temporalio/sdk-java#1793]
**Is your feature request related to a problem? Please describe.**
We would like to implement a generic interceptor that can detect that the workflow is about to fail and in this case invoke an activity (send email / post a message to kafka / alert).

**Describe the solution you'd like**
Implement `WorkflowInbouldCallsInterceptor.execute`, wrap the code in try / catch and catch the exception. In the catch block invoke our activity. Would be nice to have some static utility method that could tell us that the current exception will fail the workflow, the method should take into account `WorkflowImplementationOptions.failWorkflowExceptionTypes`.

**Describe alternatives you've considered**
The alternative is to use a base workflow class / utility class with try / catch, but still there is no generic way to tell if the current exception will fail the workflow (`WorkflowImplementationOptions.failWorkflowExceptionTypes`).

**Comment by Quinn-With-Two-Ns** [comment]
There are a few problems with this idea. 

1. The Temporal service owns workflow execution and could fail a workflow in certain situations without the worker being notified to handle it. An example would be a workflow timeout. 

2. A callback like `WorkflowInbouldCallsInterceptor.execute` cannot exist because workflow execution can take place over multiple workers during a workflow lifetime.

If we are only talking about workflow failures from exceptions then it is possible, but needs to  be done in the workflow code.

**Comment by Quinn-With-Two-Ns** [comment]
hm yes you are correcy, my apologies.

 Originally I thought this should be done in the workflow, but looks like the Java SDK does not store `failWorkflowExceptionTypes` in history so exposing in workflows or workflows interceptors exposes a potential non determinism 
 
 I wonder if you could just catch  `WorkflowExecutionException` in the interceptor?  That should be thrown if the workflow has failed.

**Comment by didvdk** [comment]
I'll check, but looking at the Temporal SDK code it looks like this exception is thrown one level above the interceptor. Also good point about non-deterministic behaviour, isn't it the case already now? If I replay a workflow with different `WorkflowImplementationOptions.failWorkflowExceptionTypes` I might get different results?

**Comment by ddavidyuk** [comment]
I'll check, but looking at the Temporal SDK code it looks like this exception is thrown one level above the interceptor. Also good point about non-deterministic behaviour, isn't it the case already now? If I replay a workflow with different WorkflowImplementationOptions.failWorkflowExceptionTypes I might get different results?

**Comment by Quinn-With-Two-Ns** [comment]
hm yes I believe you are correct. 

Since calling an activity conditional on an exception matching `failWorkflowExceptionTypes` is not deterministic I don't think we would want to add any feature that encourages this in the SDK.

If you still want to do this you can write your own interceptor or base worker and inject your `failWorkflowExceptionTypes` , but you  would need to be careful if you ever wanted to change `failWorkflowExceptionTypes` while you have workflows running

**Comment by Quinn-With-Two-Ns** [comment]
Closing as not planned since such a utility would not be deterministic. If a users really wants it they can implmenet it themselves since they control `failWorkflowExceptionTypes` and take on the risk of using it correctly.

**Comment by akshayd97** [comment]
Following up here rather than opening a new issue, since it's the same design topic. I hit this
wall too, and I think the determinism objection can be satisfied by moving the hook rather than
dropping the idea.

Agreed on the objection itself — `failWorkflowExceptionTypes` isn't in history, so it must not be
visible in workflow context. Not looking to reopen that.

The part that seems worth revisiting: #1590 (filed by @Spikhalskiy) makes the same observation about
the SDK's own metrics — that counting at the workflow-method level is wrong because it includes
replays, and that the same problem applies to `workflow_failed`. Its fix was to report *"only after reporting the
completion of the workflow task to the server and getting a confirmation that the completion was not
rejected"*, which is now `setApplyPostCompletionMetrics(...)` in `ReplayWorkflowRunTaskHandler`.

So the SDK fixed attempt-scoped counting for its own counters. User code doing the equivalent —
audit record, completion notification, span status, custom metric — is still stuck at the
workflow-method level, which is exactly the bug #1590 describes. This issue closed the only
extension point people reach for, and nothing replaced it.

Would a **worker-level** listener hung off that same post-completion point be acceptable? Outside
workflow context, after the task response is accepted, receiving the outcome the worker already
derived rather than the config it derived it from:

```java
public interface WorkflowCompletionListener {
  enum Outcome { COMPLETED, FAILED, CANCELED, CONTINUED_AS_NEW }

  void onWorkflowCompletion(CompletionInfo info);  // outcome, workflowId, runId,
                                                   // workflowType, taskQueue,
                                                   // failure, isBenign
}
```

registered on `WorkerFactoryOptions`. It emits no commands, doesn't run during replay, and gets an
immutable snapshot after the decision is made — so it can't influence determinism. Implementations
never see `failWorkflowExceptionTypes`, only its result, so the config stays worker-side.

This wouldn't cover service-owned terminal states (execution timeout, termination), as noted above —
those never reach the worker and still need Visibility. Worker-observed completions only.

The concrete gap today, on 1.32.1: with `failWorkflowExceptionTypes` set, an interceptor using the
best classification available from public API (`TemporalFailure` and not `CanceledFailure`) misses
the failure that the worker itself counts.

```
temporal_workflow_failed  = 1
interceptorTerminalCount  = 0
```

Happy to put up a PR if there's interest in the shape.
